### PR TITLE
[AppService] functionapp: Added support for in preview dotnet-isolated runtime.

### PR DESCRIPTION
**Pull request by gzuber**
**Description**

This is mainly a stacks update to add support for the new and in preview dotnet-isolated runtime. This is different from dotnet because it runs in a separate process from the functions host. I also had to add a bit of temporary custom code because of a known bug that's being fixed for creating dotnet-isolated linux consumption apps (linuxFxVersion needs to be blank in this scenario). It will be removed as soon as the bug is fixed.

**Testing Guide**

I added 3 news tests to make sure the CLI is able to create dotnet-isolated apps for a variety of scenarios (windows consumption, linux dedicated and consumption).

```
test_functionapp_on_linux_app_service_dotnet_isolated
test_functionapp_consumption_linux_dotnet_isolated
test_functionapp_windows_runtime_dotnet_isolated
```

**History Notes**

[AppService] az functionapp create: Added preview support for creating dotnet-isolated apps.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by gzuber** [comment]
@qwordy -- sorry for the late notice but is there anyway we can get this in for [S183](https://github.com/Azure/azure-cli/milestone/101)? Need this for ignite. Let me know if you need anything else from me, thanks!

**Comment by yonzhan** [comment]
@yungezz for awareness

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 1]
```diff
@@ -2972,6 +2972,10 @@ def create_function(cmd, resource_group_name, name, storage_account, plan=None,
         snake_case_prop = _convert_camel_to_snake_case(prop)
         setattr(site_config, snake_case_prop, value)
 
+    # temporary workaround for dotnet-isolated linux consumption apps
+    if is_linux and consumption_plan_location is not None and runtime == 'dotnet-isolated':
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2976]: lower() before comparasion?

Comment (gzuber) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:2976]: Yeah that's a good catch but `runtime` is actually used a lot for comparisons in this method so I just did a blanket conversion here: https://github.com/Azure/azure-cli/blob/a726dc80e02bc1ce2937f0657519ed546a11343d/src/azure-cli/azure/cli/command_modules/appservice/custom.py#L2853-L2854

**Review by qwordy — APPROVED** [review]

**Comment by gzuber** [comment]
Oh I can’t run the pipeline :/ does it run automatically or does someone have to kick it off?

**Comment by gzuber** [comment]
So I'm not really sure what's failing here... I looked into the exceptions being thrown and I'm seeing

```
2021-02-24T07:31:32.0407264Z vcr.errors.CannotOverwriteExistingCassetteException: Can't overwrite existing cassette ('/opt/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/azure/cli/command_modules/appservice/tests/latest/recordings/test_functionapp_windows_runtime_dotnet_isolated.yaml') in your current record mode ('once').
2021-02-24T07:31:32.0408927Z No match for the request (<Request (GET) https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Storage/storageAccounts/clitest000002?api-version=2021-01-01>) was found.
2021-02-24T07:31:32.0409762Z Found 1 similar requests with 1 different matcher(s) :
2021-02-24T07:31:32.0410017Z 
2021-02-24T07:31:32.0411017Z 1 - (<Request (GET) https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Storage/storageAccounts/clitest000002?api-version=2019-06-01>).
2021-02-24T07:31:32.0411992Z Matchers succeeded : ['method', 'scheme', 'host', 'port', 'path']
2021-02-24T07:31:32.0412403Z Matchers failed :
2021-02-24T07:31:32.0413008Z _custom_request_query_matcher - assertion failure :
2021-02-24T07:31:32.0413389Z None
```

But all these tests are passing locally. The weird thing is none of the recording yaml files even have a mention of `https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Storage/storageAccounts/clitest000002?api-version=2021-01-01` with that `api-version`. They all use the `api-version=2019-06-01` mentioned as a similar request found. These are also tests that were copy/pasted from older tests that pass fine with just a few values adjusted.

Is this a recent change or is there some kind of config I need to set? I tried rebasing and re-recording all the tests locally to see if that helps at all.

**Comment by gzuber** [comment]
@qwordy can you azp run this again? I may have fixed it.

I saw this commit: https://github.com/Azure/azure-cli/commit/fffde7ec2b08501805b03e6f15a17d5cb98f159f#diff-05ee69a5b7cf47cbda6c6280aefb2002174a518dc091d2bc6a899ada9145d39e

and I'm thinking I might not have been up to date when I ran the tests. I noticed all the functionapp tests now have calls with api-version=2021-01-01 after that commit so I re-ran them after rebasing and now they all have calls to that api-version

**Comment by qwordy** [comment]
> 
> 
> So I'm not really sure what's failing here... I looked into the exceptions being thrown and I'm seeing
> 
> ```
> 2021-02-24T07:31:32.0407264Z vcr.errors.CannotOverwriteExistingCassetteException: Can't overwrite existing cassette ('/opt/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/azure/cli/command_modules/appservice/tests/latest/recordings/test_functionapp_windows_runtime_dotnet_isolated.yaml') in your current record mode ('once').
> 2021-02-24T07:31:32.0408927Z No match for the request (<Request (GET) https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Storage/storageAccounts/clitest000002?api-version=2021-01-01>) was found.
> 2021-02-24T07:31:32.0409762Z Found 1 similar requests with 1 different matcher(s) :
> 2021-02-24T07:31:32.0410017Z 
> 2021-02-24T07:31:32.0411017Z 1 - (<Request (GET) https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Storage/storageAccounts/clitest000002?api-version=2019-06-01>).
> 2021-02-24T07:31:32.0411992Z Matchers succeeded : ['method', 'scheme', 'host', 'port', 'path']
> 2021-02-24T07:31:32.0412403Z Matchers failed :
> 2021-02-24T07:31:32.0413008Z _custom_request_query_matcher - assertion failure :
> 2021-02-24T07:31:32.0413389Z None
> ```
> 
> But all these tests are passing locally. The weird thing is none of the recording yaml files even have a mention of `https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Storage/storageAccounts/clitest000002?api-version=2021-01-01` with that `api-version`. They all use the `api-version=2019-06-01` mentioned as a similar request found. These are also tests that were copy/pasted from older tests (that pass fine) with just a few values adjusted.
> 
> Is this a recent change or is there some kind of config I need to set? I tried rebasing and re-recording all the tests locally to see if that helps at all.

@Zunli Hu upgraded API version of storage in Azure CLI recently. So, it can't find matching request in yaml files. Just rerun the failed test cases in --live mode. Remember to pull latest code from dev.

run "azdev test testcase --live" or set environment variable "AZURE_TEST_RUN_LIVE=True" to run live test.

**Comment by qwordy** [comment]
> 
> 
> @qwordy can you azp run this again? I may have fixed it.
> 
> I saw this commit: [fffde7e#diff-05ee69a5b7cf47cbda6c6280aefb2002174a518dc091d2bc6a899ada9145d39e](https://github.com/Azure/azure-cli/commit/fffde7ec2b08501805b03e6f15a17d5cb98f159f#diff-05ee69a5b7cf47cbda6c6280aefb2002174a518dc091d2bc6a899ada9145d39e)
> 
> and I'm thinking I might not have been up to date when I ran the tests. I noticed all the functionapp tests now have calls with api-version=2021-01-01 after that commit so I re-ran them after rebasing and now they all have calls to that api-version

@gzuber Sure. The checks should start automatically. Github is not stable.

**Comment by gzuber** [comment]
It looks like it failed `test_postgres_flexible_server_vnet_mgmt_supplied_subnetid (command_modules.rdbms.tests.latest.test_rdbms_flexible_commands.FlexibleServerVnetMgmtScenarioTest)`

I'm seeing rdbms tests failing on dev intermittently like here: 
https://github.com/Azure/azure-cli/runs/1968101738
https://github.com/Azure/azure-cli/runs/1966927311
https://github.com/Azure/azure-cli/runs/1966786703

I didn't make any changes that should affect these tests and I'm pretty close to being up to date with dev (other than commits less than a day old). Are these just flaky tests that we can get an override to merge for? Or do we have to re-run until they pass?

I went ahead and rebased on dev again to get the latest commits if we can try running again

**Comment by qwordy** [comment]
@gzuber rdbms has nothing to do with your code. I don't think it is your fault. We will have a look so that it will not block your PR.

**Comment by gzuber** [comment]
Oh good it passed! This should be good to go -- Thanks so much for all your help!

**Pull request merged**